src/dict.c

```diff
@@ -744,9 +792,19 @@ void dictSetKey(dict *d, dictEntry* de, void *key) {
         de->key = key;
 }
 
-void dictSetVal(dict *d, dictEntry *de, void *val) {
-    assert(entryHasValue(de));
-    de->v.val = d->type->valDup ? d->type->valDup(d, val) : val;
+int dictSetVal(dict *d, dictEntry **de, void *val) {
+    assert(entryHasValue(*de));
+    void *v = d->type->valDup ? d->type->valDup(d, val) : val;
+    if (entryIsEmbedded(*de)) {
+        void *key = dictGetKey(*de);
+        dictEntry *unlinked = dictUnlink(d, key);
+        *de = dictAddWithValue(d, key, val);
+        dictFreeUnlinkedEntry(d, unlinked);
+        return 1;
+    } else {
+        (*de)->v.val = v;
```

Comment:
I like the attempt at being defensive here, but I'm still concerned about how easy it will be to misuse this. If we are calling a function and pass through the pointer, someone might just do `&de`, which won't correctly update the source.